### DISCUSSION TITLE
Loading extensions in static inner classes from shaded JAR files

### DISCUSSION THREAD

**Discussion by tinkerware** [xvik/dropwizard-guicey#229] [Q&A] [answered]
Hi,

Thanks for the nicely-featured library, it's made Guice/Dropwizard integration a lot more painless for me.

I do run into an issue using class path scanning to load a Jersey extension from my app's shaded JAR. I traced it into `OReflectionHelper` filtering all inner classes out when loading them from JAR files:

https://github.com/xvik/dropwizard-guicey/blob/525575a6b4144320df6000c2fa3fa738c92045fd/src/main/java/ru/vyarus/dropwizard/guice/module/installer/scanner/util/OReflectionHelper.java#L52-L56

This doesn't really work in my case, where I use a shaded JAR for my app. Is there anything obvious I'm missing in this behavior or is this a bug in `guicey`?

Thanks,
–CKG.

**Comment by xvik** [comment, marked as answer]
Hi! 

Could you please better describe your issue: is it exception during classpath scan or inner class gets resolved as an extension or an opposite (extension not recognized)?
Small example would also be helpful (to understand what kind of path was accepted (or not))

As a general rule I always propose to use classpath scan only on classes in the current project (simply because it's faster for development). If something moved to external jar (module) it's better to register such extensions manually with a bundle (when custom bundle inclusion is possible). Still, just a recommendation, not a requirement.

Will be able to tell more with new details

**Reply by tinkerware** [reply]
> is it exception during classpath scan or inner class gets resolved as an extension or an opposite (extension not recognized)?

In my case, it is an extension (Jersey provider annotated with JAX-RS `@Provider`) that's not recognized, because it is a public static inner class.

> As a general rule I always propose to use classpath scan only on classes in the current project

I configure classpath scanning with the packages of my own app, to exclude third-party dependencies. Again, the app is deployed as a shaded JAR file, which merges the contents of all third-party dependencies into a single JAR artifact. In fact, when I run the app as exploded into class files in a directory structure, `guicey` finds my provider through classpath scanning.

Based on the [documentation on classpath scanning][1], I expected public static inner classes to be discovered. When I enable diagnostics with the system hook property, I can see from scanned class counts and recognized extensions that static inner classes in the shaded JAR file under my scan path are ignored, and my extensions are not installed.

[1]: http://xvik.github.io/dropwizard-guicey/5.0.0/guide/scan/#how-it-works

**Reply by xvik** [reply]
I reproduced the issue. Will look for the root cause.

**Reply by xvik** [reply]
Fixed. Most likely, would be released in a week or two: there should be a bugfixing-only release, but other issues are still in progress.

**Reply by xvik** [reply]
[5.6.1 released](https://github.com/xvik/dropwizard-guicey/releases/tag/5.6.1)
